616 domain knowledge facts rules should be loadable via the rest api

bnouwt opened the pull request.

Sophietje commented on smart-connector-rest-server/src/main/resources/openapi-sc.yaml:
```diff
@@ -536,6 +536,46 @@ paths:
                 $ref: '#/components/schemas/ResponseMessage'
+  /sc/knowledge:
+    post:
+      summary: Use the specified domain knowledge for the given knowledge base during activating proactive (i.e. ASk/POST) knowledge interactions. Using additional domain knowledge can increase semantic interoperability. Note that a sufficient reasoner level determines whether the domain knowledge is actually used.
```
"ASk" -> "ASK"

Sophietje reviewed (changes requested):
Two additional comments:
1. I'd recommend adding some tests to test edge cases:
   a) try to load an empty string/file
   b) try and load a file with erroneous syntax
   c) try and load a file with valid syntax but that contains something that we don't support (if that's even a thing, if so we should mention it in the docs)

2. Do we support all syntax from Apache Jena's rule syntax? I saw something about the language tag somewhere. If not, we should include what syntax is supported in the documentation

bnouwt commented on docs/docs/reasoning.md:
````diff
@@ -178,6 +178,56 @@ Now, upon receiving the new measurement the rule for the Temperature Converter w
+```
+
+:::tip
+Note that the domain fact `saref:Sensor rdfs:subClassOf saref:Device` above is represented as a body-less rule.
````
Yes, I've added clarification 👍

bnouwt commented on docs/docs/reasoning.md:
```diff
@@ -178,6 +178,56 @@ Now, upon receiving the new measurement the rule for the Temperature Converter w
+Note that the domain fact `saref:Sensor rdfs:subClassOf saref:Device` above is represented as a body-less rule.
+:::
+
+Loading the above domain knowledge into a smart connector will allow its reasoner to derive that every `saref:Sensor` in the network is also a `saref:Device` and whenever a device is requested using the graph pattern `?d rdf:type saref:Device` it will also return data from KIs with the `?s rdf:type saref:Sensor` graph pattern. The default domain knowledge for every smart connector created in a KE ...
```
Yes, level 2. I've added a tip about this with a reference to more info.

bnouwt commented on docs/docs/reasoning.md:
```diff
@@ -178,6 +178,56 @@ Now, upon receiving the new measurement the rule for the Temperature Converter w
+</TabItem>
+<TabItem value="bash" label="Rest API">
+
+When using the REST API, you can load the domain knowledge into a specific smart connector using the [`POST /sc/knowledge/`](https://github.com/TNO/knowledge-engine/blob/master/smart-connector-rest-server/src/main/resources/openapi-sc.yaml#L539) operation. The body of the request should contain `text/plain` text following the Apache Jena Rules spec mentioned above:
```
Yes, I've added a similar phrase before the Java API.

bnouwt commented on reasoner/src/main/java/eu/knowledge/engine/reasoner/util/JenaRules.java:
```diff
@@ -0,0 +1,199 @@
+import eu.knowledge.engine.reasoner.Rule;
+import eu.knowledge.engine.reasoner.api.TriplePattern;
+
+public class JenaRules {
```
Good idea, I've added short description.

bnouwt commented on docs/docs/reasoning.md:
```diff
@@ -178,6 +178,56 @@ Now, upon receiving the new measurement the rule for the Temperature Converter w
+</TabItem>
+</Tabs>
+
+
```
Agree, I've added it to the docs.

bnouwt commented on reasoner/src/main/java/eu/knowledge/engine/reasoner/util/JenaRules.java:
```diff
@@ -0,0 +1,199 @@
+	 * }
+	 * </pre>
+	 * 
+	 * Note that literals with language tags are not supported.
```
It will not happen often and is actually a characteristic of the Apache Jena Rules language, so if we say we use that specification then the language tag non-support is included. But I think adding it to the documentation is still a good idea. Done!

bnouwt commented on reasoner/src/main/java/eu/knowledge/engine/reasoner/util/JenaRules.java:
```diff
@@ -0,0 +1,199 @@
+	 */
+	public static Set<BaseRule> convertJenaToKeRules(String someRules) {
+		List<org.apache.jena.reasoner.rulesys.Rule> jenaRules = org.apache.jena.reasoner.rulesys.Rule
+				.parseRules(someRules);
```
Good catch 👍 The `ParserException` will actually propagate all the way to the [REST API](https://github.com/TNO/knowledge-engine/pull/624/files#diff-d81c65582ed725549c9323413bc56f7a9dfee1ab1ce26b950886b2b39e8eff9cR276) and is catched there to give a proper error message to the user.

The Java API does not use the Jena Rules syntax, so there we do not have to deal with this `ParserException`.

bnouwt commented on reasoner/src/main/java/eu/knowledge/engine/reasoner/util/JenaRules.java:
```diff
@@ -0,0 +1,199 @@
+			if (antecedent.isEmpty()) {
+				// create KE data rule from consequent
+
+				assert consequent.size() == 1;
```
Good point, I've tested it and if the consequent is 2 the assertion fails and throws an exception that is not catched. With assertions disabled (which is typically the case in production), the `consequent.iterator().next()` will throw a `NoSuchElementException` that also breaks the knowledge engine.

I noticed that the Apache Jena Rules syntax indeed allows multiple consequents although the antecedent is empty, so would:

```
-> (rdf:type      rdfs:range rdfs:Class) (rdfs:Resource  rdf:type  rdfs:Class).
```

and

```
-> (rdf:type      rdfs:range rdfs:Class).
-> (rdfs:Resource  rdf:type  rdfs:Class).
```

have the same semantics? In that case I can just deal with them quite easily. For now I've assumed that and changed the code to deal with multiple heads. Also added a unit test to check this.

bnouwt commented:
> Two additional comments:
> 
> 1. I'd recommend adding some tests for edge cases:
>    a) try to load an empty string/file
>    b) try and load a file with erroneous syntax
>    c) try and load a file with valid syntax but that contains something that we don't support (if that's even a thing, if so we should mention it in the docs)
> 2. Do we support all syntax from Apache Jena's rule syntax? I saw something about the language tag somewhere. If not, we should include what syntax is supported in the documentation

Thanks for the review @Sophietje! A lot of good comments. Regarding the above comments:

1a) Added this case to `eu.knowledge.engine.reasoner.api.JenaRuleConversionTest`.
1b) This is already tested in `eu.knowledge.engine.rest.api.TestDomainKnowledge.testDomainKnowledge()`.
1c) I've added tests and logic to deal with [Builtin primitives](https://jena.apache.org/documentation/inference/#RULEbuiltins) which the TKE does not support.
2) I've added a tip about language tags in the documentation 👍

bnouwt commented on reasoner/src/main/java/eu/knowledge/engine/reasoner/util/JenaRules.java:
```diff
@@ -165,7 +179,9 @@ public static Set<BaseRule> convertJenaToKeRules(String someRules) {
-		if (stringRows.size() > 0) {
+		if (stringRows.size() > 0)
+
+		{
```
👍
